delivery_location/models.py

```diff
@@ -0,0 +1,148 @@
+                                     null=False,
+                                     help_text="Business Name")
+
+    @staticmethod
+    def filter_by_field(field, field_value):
+        """Fetches list of Suppliers filtered with given field and given value.
+
+        Args:
+        field:
+            Supplier attribute.
+        field_value:
+            value which will filter the suppliers.
+
+        Returns:
+            returns a queryset containing the objects that match the specified filter conditions.
+
+        Raises:
+        FieldDoesNotExist error: if Supplier doesn't have the specified attribute.
+        """
+        if hasattr(Supplier, str(field)):
+            return Supplier.objects.filter(**{field: field_value})
+        raise FieldDoesNotExist
+
+    @staticmethod
+    def get_all_suppliers():
+        """Fetches list of all Suppliers.
+
+        Args:
+            None.
+
+        Returns:
+            returns a queryset containing all the Suppliers.
+        """
+        return Supplier.objects.all()
+
+    @staticmethod
+    def get_all_fields():
+        """Fetches list of all Supplier fields.
+
+        Args:
+            None.
+
+        Returns:
+            returns a list of all Supplier fields.
+        """
+        return ['user_name', 'first_name', 'last_name',
+                'password', 'user_type', 'business_name']
+
+
+class DeliveryLocation(models.Model):
+    user_name = models.ForeignKey(Supplier, on_delete=models.CASCADE)
+    location = models.CharField(max_length=32, validators=[MinLengthValidator(1)])
+    date = models.DateField()
+
+    def add_delivery_location(self):
+        self.save()
+        try:
+            DeliveryLocation.full_clean(self)
+        except ValidationError as e:
+            self.delete()
+            raise e
+        return self
+
+    def remove_delivery_location(self):
+        DeliveryLocation.objects.get(id=self.id).delete()
+
+    def update_date(self, newdate):
+        self.date = newdate
+        self.save()
+
+    def update_location(self, new_location):
+        if not new_location:
+            raise ValidationError("Please put a new location with length bigger than zero")
+        else:
+            self.location = new_location
+            self.save()
+
+    @staticmethod
+    def filter_by_supplier_and_location(supplier, specific_location):
+        return DeliveryLocation.objects.filter(user_name=supplier, location=specific_location)
+
+    @staticmethod
+    def filter_by_location(supplier):
+        return DeliveryLocation.objects.filter(user_name=supplier)
+
+    @staticmethod
+    def get_all_delivery_locations():
+        return DeliveryLocation.objects.all()
+
+    @staticmethod
+    def get_delivery_location_by_id(a_id):
+        return DeliveryLocation.objects.get(id=a_id)
+
+    @staticmethod
+    def remove_delivery_location_by_id(a_id):
+        DeliveryLocation.objects.get(id=a_id).delete()
```

Comment:
I see here a bunch of meaningless wrapper methods... and a bunch of meaningless tests for all of them. those are Django inner methods, we don't need to rewrite or test them - The people working on Django have already done it. Please don't waste time on it.

Reply:
Regular delete method does not raise any doesnotexist exception it only raises protected exception but in our case we didnt use any protected fields all of them cascade,since i want to get an exception if specific id doesnt exist so i use the get method that raises Object(DilverLocation)DoesNotExist exception

Reply:
There's still a race condition here, BTW. It's still possible that the object will be there by the time you fetch it using get and will not be there by the time the delete member method is called.

I agree that having those wrappers is a bit redundant and generates additional maintenance effort for which I still don't see the benefit.

Reply:
When a supplier deletes a dilvery location i want the supplier to know that he deleting something that exist other wise he gets a meesage like exception,regular delete method does not rise exception even if the object we are trying to delete is not there in the database,so i want something that can warn us that we are deleting something that does not exist.

Reply:
I see the point, and this is why I was ok with that on other modules as well.

Having said that, this logic still contains the race condition I mentioned above, which practically brings us to square one, so it's not a deal-breaker for me, but it's something that you should be aware of if you try to further build logic on top of this later on (I can't see at the moment how it might get you troubles -- but keep it in mind).

Reply:
Are you talking about the situation when we have two fetches one for get the object and one for delete the object,so when the delete method called it stops at the get method and the second fetch get the object that has been deleted(but not deleted from the database yet)?

Reply:
As the code is written now, you're ultimately using the member's `delete()` method, you're just making sure it exists in the database by fetching the object from the database before deleting it.

What I'm saying is that it might be that the object will be deleted (by someone else) between the time you fetched the object (`DeliveryLocation.objects.get(id=a_id)`) and the time you call its `delete()` method (`.delete()`).

Again, I'm not seeing any problem with that at the moment, but just keep this race condition in mind.

Reply:
Yes since it is deleted by someone else and i am trying to delete it by the time it is deleted it is fine because no exception will be raised(we will be doing that only the supplier that has the deliverylocation can actually delete his stuff in the future)

Reply:
I hijacked the thread a bit, but I think @Omeramsc's comment was about some of the other static methods as well.

Reply:
What other static methods,all has been done in the same logic with tests

Reply:
I still don't see how this method is needed. where is the change in the exception coming from? all we're doing is calling `get` and `delete`, we can do the exact same thing without this function, yielding the same result. there is no special exception handling or a use of full clean here.

Reply:
We are using two functions that are written in one,so in order to use thos two functions we call the static function that has a meaning of what we are doing instead of just code and in the future this function can be written differently in order to write that abstract two functions in views handling we just write the static method that is already writted with more logic

Reply:
I don't fully agree with this solution, but I understand what you were going for. I'll approve it.